### PR TITLE
Add support for BL2 in XIP memory

### PR DESCRIPTION
**Pull request by JiafeiPan**
When bl2 runs in el3, it will be the first TF image to
be executed, so in some cases bl2 will be in ROM just
like bl1. This patch enable bl2 in ROM in case of it
runs in el3.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

### CODE REVIEW

**Comment by antonio-nino-diaz-arm** [comment]
Can't the current linkerscript be modified with some ifdefs with `BL2_AT_EL3_ROM`? I don't like the idea of having 2 different files that are basically the same with some small differences...

**Comment by JiafeiPan** [comment]
Hi, Antonio,
Thanks for your comments, I have updated the patch according to your comments. please review again.

Best Regards,
Jiafei.

**Comment by antonio-nino-diaz-arm** [comment]
This looks good to me, but someone else should take a look.

**Comment by antonio-nino-diaz-arm** [comment]
IP Review +1

**Comment by JiafeiPan** [comment]
Thanks Antonio, have updated copyright years.

**Comment by danh-arm** [comment]
I don't understand the need for this patch. If you want to run in ROM, you should use BL1. If there is some feature of BL2 you want in BL1, we should be looking to move that functionality there. What is it in BL2 that you need?

I know there are some differences between the image loading behaviour in BL1 and BL2; is that the rationale for this?

**Comment by JiafeiPan** [comment]
Hi, danh-arm, thanks for your comments. Layerscape already has an internal bootrom, and it support Nor flash boot and SD, usb and the other boot. For SD, usb and all the other boot, internal bootrom will load the first TF-A images into RAM and then execute it in RAM, so we can use BL2 as the first TF image which runs in RAM. But for nor flash boot, internal boot rom will not load the TF-A imges into RAM and will execute it in nor flash directly. Currently we use BL1 as the first TF-A images, but if BL2 can run in the ROM,  we can also remove BL1 and use BL2 as the first image which runs in ROM directly, so that we can save more internal RAM for the following booting stage. This is our use case which need to run BL2 in ROM. Any comments is welcome, thanks.

**Comment by danh-arm** [comment]
Hi @JiafeiPan.

Thanks for the explanation. I understand your use-case now and it makes sense. However, I think it's confusing to use the term "ROM" in this use-case, because if you were able to program this as a real ROM in the factory, then you would also be able use TF BL1 as your real ROM, instead of your existing internal bootrom. I think it's better to use the term "eXecute In Place (XIP) memory" instead, to make it clear that BL2 is stored in XIP (but updateable) memory.

Other comments to your patch will follow...

Regards

Dan.

File: Makefile [hunk 1]
```diff
@@ -518,6 +518,7 @@ $(eval $(call assert_boolean,USE_COHERENT_MEM))
 $(eval $(call assert_boolean,USE_TBBR_DEFS))
 $(eval $(call assert_boolean,WARMBOOT_ENABLE_DCACHE_EARLY))
 $(eval $(call assert_boolean,BL2_AT_EL3))
+$(eval $(call assert_boolean,BL2_AT_EL3_ROM))
```
Comment (danh-arm) [inline Makefile]: I think the new flag should be called BL2_IN_XIP_MEM. The term ROM is confusing in this context. Also, this feature potentially could be used when BL2 is in Secure-EL1, so this should be orthgonal to the BL2_AT_EL3 feature. Having said that, for now you could avoid supporting the case where BL2_AT_EL3 = 0 and BL2_IN_XIP_MEM = 1, by throwing an error in the top level Makefile (in the "Check incompatible options" section).

Given the above, I think your commit message should look something like this:

Add support for BL2 in XIP memory

In some use-cases BL2 will be stored in eXecute In Place (XIP) memory, like BL1. In these use-cases, it is necessary to initialize the RW sections in RAM, while leaving the RO sections in place. This patch enable this use-case with a new build option, BL2_IN_XIP_MEM. For now, this option is only supported when BL2_AT_EL3 is 1.

File: bl2/bl2_el3.ld.S [hunk 2]
```diff
@@ -12,15 +12,26 @@ OUTPUT_ARCH(PLATFORM_LINKER_ARCH)
 ENTRY(bl2_entrypoint)
 
 MEMORY {
+#if BL2_AT_EL3_ROM
+    ROM (rx): ORIGIN = BL2_RO_BASE, LENGTH = BL2_RO_LIMIT - BL2_RO_BASE
+    RAM (rwx): ORIGIN = BL2_RW_BASE, LENGTH = BL2_RW_LIMIT - BL2_RW_BASE
```
Comment (danh-arm) [inline bl2/bl2_el3.ld.S]: The porting guide will need updating to describe these new platform porting constants, and to say that these are only needed when BL2_IN_XIP_MEM is 1. Also, to say that BL2_LIMIT and BL2_BASE are not applicable in this case.

File: bl2/bl2_el3.ld.S [hunk 3]
```diff
@@ -174,12 +207,32 @@ SECTIONS
...
+#if BL2_AT_EL3_ROM
+    __BL2_RAM_START__ = ADDR(.data);
+    __BL2_RAM_END__ = .;
+
+    __DATA_ROM_START__ = LOADADDR(.data);
+    __DATA_SIZE__ = SIZEOF(.data);
+
+    /*
+     * The .data section is the last PROGBITS section so its end marks the end
+     * of BL2's actual content in Trusted ROM.
```
Comment (danh-arm) [inline bl2/bl2_el3.ld.S]: minor: Perhaps say "... of BL2's RO content in XIP memory." here.

File: bl2/bl2_el3.ld.S [hunk 4]
```diff
@@ -174,12 +207,32 @@ SECTIONS
...
+     */
+    __BL2_ROM_END__ =  __DATA_ROM_START__ + __DATA_SIZE__;
+    ASSERT(__BL2_ROM_END__ <= BL2_RO_LIMIT,
+           "BL2's ROM content has exceeded its limit.")
```
Comment (danh-arm) [inline bl2/bl2_el3.ld.S]: minor: s/ROM/RO/

File: bl2/bl2_el3.ld.S [hunk 5]
```diff
@@ -174,12 +207,32 @@ SECTIONS
...
     __BSS_SIZE__ = SIZEOF(.bss);
 
+
 #if USE_COHERENT_MEM
     __COHERENT_RAM_UNALIGNED_SIZE__ =
         __COHERENT_RAM_END_UNALIGNED__ - __COHERENT_RAM_START__;
 #endif
 
+#if BL2_AT_EL3_ROM
+    ASSERT(. <= BL2_RW_LIMIT, "BL2 image has exceeded its limit.")
```
Comment (danh-arm) [inline bl2/bl2_el3.ld.S]: minor: "BL2's RW content has ...."

File: docs/user-guide.rst [hunk 6]
```diff
@@ -246,6 +246,9 @@ Common build options
 - ``BL2_AT_EL3``: This is an optional build option that enables the use of
    BL2 at EL3 execution level.
 
+- ``BL2_AT_EL3_ROM``: This is an optional build option that enables the use of
+   BL2 at EL3 execution level, and BL2 read only section is in ROM.
```
Comment (danh-arm) [inline docs/user-guide.rst]: Will need rewording like above.

File: include/common/aarch64/el3_common_macros.S [hunk 7]
```diff
@@ -278,8 +278,8 @@
 		 * an earlier boot loader stage.
 		 * -------------------------------------------------------------
 		 */
-		adr	x0, __RW_START__
-		adr	x1, __RW_END__
+		ldr	x0, =__RW_START__
+		ldr	x1, =__RW_END__
```
Comment (danh-arm) [inline include/common/aarch64/el3_common_macros.S]: What's the rationale for this change? Using adr is preferred where possible. If this really is needed, it will need to be in a separate commit with explanation.

File: Makefile (hunk 1 above)
Comment (JiafeiPan) [reply Makefile]: Great, will modify it, thanks.

File: bl2/bl2_el3.ld.S (hunk 2 above)
Comment (JiafeiPan) [reply bl2/bl2_el3.ld.S]: Done, thanks.

File: bl2/bl2_el3.ld.S (hunk 3 above)
Comment (JiafeiPan) [reply bl2/bl2_el3.ld.S]: Done, thanks.

File: bl2/bl2_el3.ld.S (hunk 4 above)
Comment (JiafeiPan) [reply bl2/bl2_el3.ld.S]: Done, thanks.

File: bl2/bl2_el3.ld.S (hunk 5 above)
Comment (JiafeiPan) [reply bl2/bl2_el3.ld.S]: Done, thanks.

File: docs/user-guide.rst (hunk 6 above)
Comment (JiafeiPan) [reply docs/user-guide.rst]: Done, thanks.

File: include/common/aarch64/el3_common_macros.S (hunk 7 above)
Comment (JiafeiPan) [reply include/common/aarch64/el3_common_macros.S]: For the adr instruction, it require the lable's offset from the address of this instruction must be in the range +/-1MB. If the option "BL2_IN_XIP_MEM" is set to '1', in some cases, BL2's RW memory will not in the range of +/-1MB from BL2's RO memory region. so we need to use ldr instruction to cover this case.

**Comment by JiafeiPan** [comment]
Hi, Dan, have updated the patches according to your comments, please review again, thanks.

File: include/common/aarch64/el3_common_macros.S (hunk 7 above)
Comment (danh-arm) [reply include/common/aarch64/el3_common_macros.S]: OK, makes sense - thanks.
minor: Could you just fix the spelling in the commit message (limiation and lable)?

File: Makefile [hunk 8]
```diff
@@ -564,6 +565,7 @@ $(eval $(call add_define,USE_COHERENT_MEM))
 $(eval $(call add_define,USE_TBBR_DEFS))
 $(eval $(call add_define,WARMBOOT_ENABLE_DCACHE_EARLY))
 $(eval $(call add_define,BL2_AT_EL3))
+$(eval $(call add_define,BL2_IN_XIP_MEM))
```
Comment (danh-arm) [inline Makefile]: I don't see the check I suggested in my last review: 
"... you could avoid supporting the case where BL2_AT_EL3 = 0 and BL2_IN_XIP_MEM = 1, by throwing an error in the top level Makefile (in the "Check incompatible options" section)"

Comment (JiafeiPan) [reply Makefile]: Sorry, I thought that we need to add build check in platform makefile, will add it to top Makefile, thanks.

Comment (JiafeiPan) [reply Makefile]: Hi, Dan, have updated the patch according to your comments, please review again, thanks.

**Comment by danh-arm** [comment]
jenkins: test this please

**Comment by dp-arm** [comment]
@JiafeiPan can you resolve the conflicts and then I can merge?

**Comment by JiafeiPan** [comment]
Hi, @dp-arm, have resolved the conflicts, please merge it, thanks.
